sql/core/src/test/scala/org/apache/spark/sql/execution/OptimizeMetadataOnlySuite.scala

```diff
@@ -0,0 +1,95 @@
+import org.apache.spark.sql.test.SharedSQLContext
+
+class OptimizeMetadataOnlySuite extends QueryTest with SharedSQLContext {
+  import testImplicits._
+
+  override def beforeAll(): Unit = {
+    super.beforeAll()
+    val data = (1 to 10).map(i => (i, s"data-$i", i % 2, if ((i % 2) == 0) "even" else "odd"))
+      .toDF("id", "data", "partId", "part")
+    data.write.partitionBy("partId", "part").mode("append").saveAsTable("srcpart_15752")
+  }
+
+  override protected def afterAll(): Unit = {
+    try {
+      sql("DROP TABLE IF EXISTS srcpart_15752")
+    } finally {
+      super.afterAll()
+    }
+  }
+
+  private def checkWithMetadataOnly(df: DataFrame): Unit = {
+    val localRelations = df.queryExecution.optimizedPlan.collect {
+      case l @ LocalRelation(_, _) => l
+    }
+    assert(localRelations.size == 1)
+  }
+
+  private def checkWithoutMetadataOnly(df: DataFrame): Unit = {
+    val localRelations = df.queryExecution.optimizedPlan.collect{
+      case l @ LocalRelation(_, _) => l
+    }
+    assert(localRelations.size == 0)
+  }
+
+  test("spark-15752 metadata only optimizer for partition table") {
+    withSQLConf(SQLConf.OPTIMIZER_METADATA_ONLY.key -> "true") {
+      checkWithMetadataOnly(sql("select part from srcpart_15752 where part = 0 group by part"))
+      checkWithMetadataOnly(sql("select max(part) from srcpart_15752"))
+      checkWithMetadataOnly(sql("select max(part) from srcpart_15752 where part = 0"))
+      checkWithMetadataOnly(
+        sql("select part, min(partId) from srcpart_15752 where part = 0 group by part"))
+      checkWithMetadataOnly(
+        sql("select max(x) from (select part + 1 as x from srcpart_15752 where part = 1) t"))
+      checkWithMetadataOnly(sql("select distinct part from srcpart_15752"))
+      checkWithMetadataOnly(sql("select distinct part, partId from srcpart_15752"))
+      checkWithMetadataOnly(
+        sql("select distinct x from (select part + 1 as x from srcpart_15752 where part = 0) t"))
+
+      // Now donot support metadata only optimizer
+      checkWithoutMetadataOnly(sql("select part, max(id) from srcpart_15752 group by part"))
+      checkWithoutMetadataOnly(sql("select distinct part, id from srcpart_15752"))
+      checkWithoutMetadataOnly(sql("select part, sum(partId) from srcpart_15752 group by part"))
+      checkWithoutMetadataOnly(
+        sql("select part from srcpart_15752 where part = 1 group by rollup(part)"))
+      checkWithoutMetadataOnly(
+        sql("select part from (select part from srcpart_15752 where part = 0 union all " +
+          "select part from srcpart_15752 where part= 1)t group by part"))
+    }
+  }
+
+  test("spark-15752 without metadata only optimizer for partition table") {
+    withSQLConf(SQLConf.OPTIMIZER_METADATA_ONLY.key -> "false") {
+      checkWithoutMetadataOnly(sql("select part from srcpart_15752 where part = 0 group by part"))
+      checkWithoutMetadataOnly(sql("select max(part) from srcpart_15752"))
```

Comment:
ah ok - this is getting confusing. we should just have checkWithMetadataOnly test the behavior when the flag is off.